pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -279,18 +293,26 @@ func (rc *reconciler) attachDesiredVolumes() {
 			}
 		}
 
+		// insert a detach operation start time into cache if needed for the volume for metrics reporting
+		// note that the plugin name has been set to "" as there is no visibility at this moment to which
+		// plugin/csi driver would be conducting the operation. It will be populated later when detach
+		// operation function is created and plugin name then could be decided
+		rc.operationStartTimeCache.AddIfNotExist(createOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName), "", "volume_attach")
```

Comment:
This does not cover end to end operational metrics as it should. I would expect, end to end metrics to cover time from which attach was request to time when attach was finished. This metric does not include time for:

1. Volume to detach from previous node if an operation was pending
2. Time waiting for reconciler to kick the operation in.

As such - I think the numbers reported here will be very close to already pre-existing operation metrics and hence aren't very useful as brand new metrics.

Reply:
for
1. my understanding is there are two cases:
    a. there is an attach operation pending to be finished, which is fine not to add anything into the cache
    b. there is a detach operation needs to happen before the attach can be kicked off. There should be a detach metric being recorded in this case, and getting ride of that piece from the whole attach operation process seems to be also reasonable to me, the end of day the metric reports end to end latency of attach operation.
  2. "Time waiting for reconciler to kick the operation in" I assume you meant the time elapsed until "attachDesiredVolumes" will be called? Now the question will still be will that new metric should record controller related delay or not. If not, ignoring the time elapsed before an actual attach operation is valid.

for the last piece, It will only be close to already existed operation metrics if there is NO error happened during the whole process. If there's any error, the real start time of attach operation will be gone along with goroutine finishes "GenerateAttachVolumeFunc" returned "onComplete" call, which is a major fix of this PR

Reply:
discussed offline. The metric will not include the waiting time when there's a pending operation on a specific volume for the sake of eliminating confusions and to simply SLA/alert set up for end users